benchmarks/graphviz/graphviz-2.34.0/lib/graph/graph.3
.TH LIBGRAPH 3 "01 MARCH 1993"
.SH NAME
\fBlibgraph\fR \- abstract graph library
.SH SYNOPSIS
.ta .75i 1.5i 2.25i 3i 3.75i 4.5i 5.25i 6i
.PP
.nf
\f5
#include <graphviz/graph.h>
void        aginit();
Agraph_t    *agread(FILE*);
int         agwrite(Agraph_t*, FILE*);
int         agerrors();
Agraph_t    *agopen(char *name, int kind);
void        agclose(Agraph_t *g);
Agraph_t    *agsubg(Agraph_t *g, char *name);
Agraph_t    *agfindsubg(Agraph_t *g, char *name);
Agnode_t    *agmetanode(Agraph_t *g);
Agraph_t    *agusergraph(Agnode_t *metanode);
int         agnnodes(Agraph_t *g), agnedges(Agraph_t *g);
.sp .i1
int         agcontains(Agraph_t *g, void *obj);
int         aginsert(Agraph_t *g, void *obj);
int         agdelete(Agraph_t *g, void *obj);
.sp .1i
Agnode_t    *agnode(Agraph_t *g, char *name);
Agnode_t    *agfindnode(Agraph_t *g, char *name);
Agnode_t    *agfstnode(Agraph_t *g);
Agnode_t    *agnxtnode(Agraph_t *g, Agnode_t *n);
Agnode_t    *aglstnode(Agraph_t *g);
Agnode_t    *agprvnode(Agraph_t *g, Agnode_t *n);
.sp .1i
Agedge_t    *agedge(Agraph_t *g, Agnode_t *tail, Agnode_t *head);
Agedge_t    *agfindedge(Agraph_t *g, Agnode_t *tail, Agnode_t *head);
Agedge_t    *agfstedge(Agraph_t *g, Agnode_t *n);
Agedge_t    *agnxtedge(Agraph_t *g, Agedge_t *e, Agnode_t *n);
Agedge_t    *agfstin(Agraph_t *g, Agnode_t *n);
Agedge_t    *agnxtin(Agraph_t *g, Agedge_t *e);
Agedge_t    *agfstout(Agraph_t *g, Agnode_t *n);
Agedge_t    *agnxtout(Agraph_t *g, Agedge_t *e);
.sp .1i
char        *agget(void *obj, char *name);
char        *agxget(void *obj, int index);
void        agset(void *obj, char *name, char *value);
void        agxset(void *obj, int index, char *value);
int         agindex(void *obj, char *name);
.sp .1i
Agsym_t*    agraphattr(Agraph_t *g,char *name,char *value);
Agsym_t*    agnodeattr(Agraph_t *g,char *name,char *value);
Agsym_t*    agedgeattr(Agraph_t *g,char *name,char *value);
Agsym_t*    agfindattr(void *obj,char *name);
\fP
.fi
.SH DESCRIPTION
\fIlibgraph\fP maintains directed and undirected attributed graphs
in memory and reads and writes graph files.  Graphs are composed of
nodes, edges, and nested subgraphs.   A subgraph may contain any
nodes and edges of its parents, and may be passed to any
\fIlibgraph\fP function taking a graph pointer, except the three
that create new attributes (where a main graph is required).

Attributes are internal or external.
Internal attributes are fields in the graph, node and edge structs
defined at compile time.
These allow efficient representation and direct access to values
such as marks, weights, and pointers for writing graph algorithms.
External attributes, on the other hand, are character strings
(name\(hyvalue pairs) dynamically allocated at runtime and accessed
through \fIlibgraph\fP calls.  External attributes are used in
graph file I/O; internal attributes are not.  Conversion between
internal and external attributes must be explicitly programmed.

The subgraphs in a main graph are represented by an auxiliary directed
graph (a meta\(hygraph).  Meta\(hynodes correspond to subgraphs, and meta\(hyedges
signify containment of one subgraph in another. 
\f5agmetanode\fP and \f5agusergraph\fP map between
subgraphs and meta\(hynodes.  The nodes and edges of the meta\(hygraph may
be traversed by the usual \fIlibgraph\fP functions for this purpose.

.SH USE
1. Define types \f5Agraphinfo_t\fP, \f5Agnodeinfo_t\fP,
and \f5Agedgeinfo_t\fP (usually in a header file) before
including \f5<graphviz/graph.h>\fP. 

2. Call \f5aginit()\fP before any other \fIlibgraph\fP functions.
(This is a macro that calls \f5aginitlib()\fP to define the sizes
of Agraphinfo_t, Agnodeinfo_t, and Agedgeinfo_t.)

3. Compile with \-lgraph \-lcdt.

Except for the \fBu\fP fields, \fIlibgraph\fP
data structures must be considered read\(hyonly.
Corrupting their contents by direct updates can cause
catastrophic errors.

Warning: the library is not thread-safe.

.SH "GRAPHS"
.nf
\f5
typedef struct Agraph_t {
    char                 kind;
    char                *name;
    Agraph_t             *root;
    char                **attr;
    graphdata_t         *univ;
    Dict_t              *nodes,*inedges,*outedges;
    proto_t             *proto;
    Agraphinfo_t         u;
} Agraph_t;

typedef struct graphdata_t {
    Dict_t              *node_dict;
    attrdict_t          *nodeattr, *edgeattr, *globattr;
} graphdata_t;

typedef struct proto_t {
    Agnode_t            *n;
    Agedge_t            *e;
    proto_t             *prev;
} proto_t;
\fP
.fi
A graph \fIkind\fP is one of:
AGRAPH, AGRAPHSTRICT, AGDIGRAPH, or AGDIGRAPHSTRICT.
There are related macros for testing the properties of a graph:
AG_IS_DIRECTED(g) and AG_IS_STRICT(g).
Strict graphs cannot have self\(hyarcs or multi\(hyedges.
\fBattr\fP is the array of external attribute values.
\fBuniv\fP points to values shared by all subgraphs of a main graph.
\fBnodes\fP, \fBinedges\fP, and \fBoutedges\fP are sets maintained
by \fBcdt(3)\fP.  Normally you don't access these dictionaries
directly, though the edge dictionaries may be re\(hyordered to support
programmer\(hydefined ordered edges (see \f5dtreorder\fP in \fIcdt(3)\fP).
\fBproto\fP is a stack of templates for node and edge initialization.
The attributes of these nodes and edges are set in the usual way (\f5agget\fP,
\f5agset\fP, etc.) to set defaults.
.PP
\f5agread\fP reads a file and returns a new graph if one
was successfully parsed, otherwise returns NULL if
\f5EOF\fP or a syntax error was encountered.
Errors are reported on stderr and a count is returned from
\g5agerrors()\fP.
\f5write_graph\fP prints a graph on a file.
\f5agopen\fP and \f5agsubg\fP create new empty graph and subgraphs.
\f5agfindsubg\fP searches for a subgraph by name, returning NULL
when the search fails.

.SH ALL OBJECTS
\f5agcontains\fP, \f5aginsert\fP, \f5agdelete\fP are generic functions
for nodes, edges, and graphs.  \f5gcontains\fP is a predicate that tests
if an object belongs to the given graph.  \f5aginsert\fP inserts an
object in a graph and \f5agdelete\fP undoes this operation.
A node or edge is destroyed (and its storage freed) at the time it
is deleted from the main graph.  Likewise a subgraph is destroyed
when it is deleted from its last parent or when its last parent is deleted.

.SH NODES
.nf
\f5
typedef struct Agnode_t {
    char                *name;
    Agraph_t            *graph;
    char                **attr;
    Agnodeinfo_t        u;
} Agnode_t;
\fP
.fi

\f5agnode\fP attempts to create a node.
If one with the requested name already exists, the old node 
is returned unmodified.
Otherwise a new node is created, with attributed copied from g\->proto\->n.
\f5agfstnode\fP (\f5agnxtnode\fP) return the first (next) element
in the node set of a graph, respectively, or NULL.
\f5aglstnode\fP (\f5agprvnode\fP) return the last (previous) element
in the node set of a graph, respectively, or NULL.

.SH EDGES
.nf
\f5
typedef struct Agedge_t {
    Agnode_t            *head,*tail;
    char                **attr;
    Agedgeinfo_t        u;
} Agedge_t;
\fP
.fi
\f5agedge\fP creates a new edge with the attributes of g\->proto\->e
including its key if not empty.
\f5agfindedge\fP finds the first (u,v) edge in \f5g\fP.
\f5agfstedge\fP (\f5agnxtedge\fP) return the first (next) element
in the edge set of a graph, respectively, or NULL.
\f5agfstin\fP, \f5agnxtin\fP, \f5agfstout\fP, \f5agnxtout\fP
refer to in\(hy or out\(hyedge sets.
The idiomatic usage in a directed graph is:
.sp
\f5    for (e = agfstout(g,n); e; e = agnextout(g,e)) your_fun(e);\fP
.P
An edge is uniquely identified by its endpoints and its \f5key\fP
attribute (if there are multiple edges).
If the \f5key\fP of \f5g\->proto\->e\fP is empty,
new edges are assigned an internal value.
Edges also have \f5tailport\fP and \f5headport\fP values.
These have special syntax in the graph file language but are
not otherwise interpreted.
.PP
.SH ATTRIBUTES
.nf
\f5
typedef struct attrsym_t {
    char                *name,*value;
    int                 index;
    unsigned char       printed;
} attrsym_t;
.bp
typedef struct attrdict_t  {
    char                *name;
    Dict_t              *dict;
    attrsym_t           **list;
} attrdict_t;
\fP
.fi
\f5agraphattr\fP, \f5agnodeattr\fP, and \f5agedgeattr\fP make new attributes.
\f5g\fP should be a main graph, or \f5NULL\fP for declarations
applying to all graphs subsequently read or created.
\f5agfindattr\fP searches for an existing attribute.
.PP
External attributes are accessed by \f5agget\fP and \f5agset\fP
These take a pointer to any graph, node, or edge, and an attribute name.
Also, each attribute has an integer index.  For efficiency this index
may be passed instead of the name, by calling \f5agxget\fP and \f5agxset\fP.
The \f5printed\fP flag of an attribute may be set to 0 to skip it
when writing a graph file.
.PP
The \f5list\fP in an attribute dictionary is maintained in order of creation
and is NULL terminated.
Here is a program fragment to print node attribute names:
.nf
    \f5attrsym_t *aptr;
    for (i = 0; aptr = g\->univ\->nodedict\->list[i]; i++) puts(aptr\->name);\fP
.fi
.SH EXAMPLE GRAPH FILES
.nf
graph any_name {            /* an undirected graph */
    a \-\- b;                 /* a simple edge */
    a \-\- x1 \-\- x2 \-\- x3;    /* a chain of edges */
    "x3.a!" \-\- a;           /* quotes protect special characters */
    b \-\- {q r s t};         /* edges that fan out */
    b [color="red",size=".5,.5"];   /* set various node attributes */
    node [color=blue];      /* set default attributes */
    b \-\- c [weight=25];     /* set edge attributes */
    subgraph sink_nodes {a b c};    /* make a subgraph */
}

digraph G {
    size="8.5,11";            /* sets a graph attribute */
    a \-> b;                 /* makes a directed edge */
    chip12.pin1 \-> chip28.pin3; /* uses named node "ports" */
}
.fi

.SH SEE ALSO
.BR dot (1),
.BR neato (1),
.BR libdict (3)
.br
S. C. North and K. P. Vo, "Dictionary and Graph Libraries''
1993 Winter USENIX Conference Proceedings, pp. 1\(hy11.

.SH AUTHOR
Stephen North (north@ulysses.att.com), AT&T Bell Laboratories.


NEWS
KLEE 1.4.0, 21 July 2017
========================
(Incorporating changes from 4 November 2016 up to and including 21 July 2017)
Documentation at http://klee.github.io/releases/docs/v1.4.0/

This will be the last version supporting LLVM 2.9 and the autoconf build system.

- New CMake build system (@delcypher, @jirislaby)
- Added support for vectorized instructions (@delcypher)
- Fixed and documents BFS searcher behaviour (@MartinNowack, @ccadar)
- Renames .pc files to .kquery files (@holycrap872)
- Removed option --randomize-fork (@ccadar)
- Changed preferred permissions from 0622 to the more standard 0644 in the
POSIX model (@ccadar)
- New target name, "make systemtests", for running the system tests.  This replaces "make test".  Running the unit tests is still accomplished via "make unittests".
- Better support for MacOS (@andreamattavelli, @delcypher)
- Enabled support for ASan builds of KLEE (@delcypher)
- Support long long values in --stop-after-n-instructions for LLVM > 2.9
(@andreamattavelli)
- Teach KLEE to respect the requested memory alignment of globals and stack
variables when possible (@delcypher)
- Added new option --warnings-only-to-file which causes warnings to be written
to warnings.txt only (@ccadar)
- metaSMT improvements (@hoangmle)
- KLEE-web improvements (@andronat, @helicopter88)
- Fixed bug in the implementation of NotExpr (@delcypher)
- Fixed a bug leading to data loss when writing into files (@ccadar, @gladtbx)
- Some improvements and refactoring to the Expr library (@delcypher)
- Added missing constant folding opportunity when handling constant arrays
(@andreamattavelli, @delcypher)
- Teach klee::getDirectCallTarget() to resolve weak aliases (@delcypher)
- Fixed handling of internal forks (@gladtbx)
- Improved replay using libkleeRuntest (@delcypher)
- Added -debug-assignment-validating-solver feature to check the correctness
of generated assignments (@delcypher)
- Added -debug-z3-dump-queries, -debug-z3-validate-models and
-debug-z3-verbosity options useful for debugging the interaction with Z3
(@delcypher)
- Added geq/lt-llvm- configs in lit (@jirislaby)
- Work on supporting newer LLVM versions (@jirislaby)
- Fixed bug where stats would not be updated on early exit (@delcypher)
- Reworked the external dispatching mechanism (@delcypher)
- Added support for creating release documentation (@delcypher)
- Smaller refactorings, fixes and improvements, test cases, maintenance,
comments, website, etc. (@adrianherrera, @akshaynagpal, @AlexxNica,
@andreamattavelli, @bananaappletw, @bigelephant29, @bunseokbot, @ccadar,
@delcypher, @emlai, @hoangmle, @jirislaby, @kren1, @levex,
@Manouchehri, @MartinNowack, @mechtaev, @Mic92, @omeranson, @rtrembecky,
@thestr4ng3r, @tomek-kuchta)


KLEE 1.3.0, 30 November 2016
============================
(Incorporating changes from 1 April up to and including 3 November 2016)

* Improved determinism of KLEE, an essential feature for experiments involving KLEE (@MartinNowack)
* KLEE-web has been improved and refactored, and now available at http://klee.doc.ic.ac.uk/ (@giacomoguerci, @helicopter88, @andronat, @ccadar, based on work by @ains, @ben-chin, @ilovepjs, @JamesDavidCarr, Kaho Sato, Conrad Watt, @ccadar)
* Renamed --replay-out to --replay-ktest and --replay-out-dir to replay-ktest-dir (@delcypher)
* Split creation of symbolic files and stdin in two distinct options, documented at http://klee.github.io/docs/options/#symbolic-environment (@andreamattavelli)
* Support for logging queries before invoking the solver via --log-partial-queries-early, useful for debugging solver crashes (@MartinNowack)
* Added --stats-write-after-instructions and --istats-write-after-instructions to update each statistic after n steps (@MartinNowack)
* Added --compress-log and --debug-compress-instructions to gzip-compress logs (@MartinNowack)
* Added --exit-on-error-type option for stopping execution when certain error types are encountered (@jirislaby)
* Updated and improved metaSMT support and added TravisCI targets (@hoangmle)
* Added option --debug-crosscheck-core-solver to allow crosschecking of solvers (@delcypher)
* Explicitly made division total in STP (@ccadar)
* Extended support for assembler raising (@MartinNowack)
* Disabled --solver-optimize-divides, as the optimization is currently buggy (@ccadar)
* Improved --debug-print-instructions options with more logging options (@andreamattavelli)
* Improved stub for times() not to trigger a NULL dereference (@ccadar)
* Allow relocation of installed KLEE tree (@ShayDamir)
* Fixed bug in independent solver (@delcypher)
* Fixed alignement of varargs (@MartinNowack)
* Fixed variable shifting behavior with different sizes and generation of STP shift operations with variable amounts (@MartinNowack)
* Fixed handling of non-sized globals (@jirislaby)
* Fixed klee_get_obj_size() crash on 64-bit (@hutoTUM)
* Fixed bug in Kleaver's parser (@andreamattavelli)
* Refactorings, small fixes and improvements, test cases, maintenance and website: (@andreamattavelli, @ccadar, @delcypher, @domainexpert, @giacomoguerci, @hoangmle, @helicopter88, @jirislaby, @Justme0, @kren1, @MartinNowack, @mchalupa)


KLEE 1.2.0, 31 March 2016
=========================

* Added native support for Z3 (@delcypher)
* Made it possible to build KLEE without using STP and only MetaSMT (@delcypher)
* Added support for tcmalloc, which allows KLEE to better track memory consumption (@MartinNowack)
* Added support for lowering the ``llvm.objectsize`` intrinsic (@delcypher)
* Added soname for Runtest dynamic library (@MartinNowack)
* Added support to load libraries from command line (@omeranson)
* Added command line flag --silent-klee-assume to suppress errors due to infeasible assumptions (Valentin Wüstholz, @wuestholz)
* Changed code to print out arrays deterministically (@MartinNowack)
* Improved klee-clang script (@msoos)
* Added code to dump queries and assignments (@MartinNowack)
* Code cleanup and refactorings (@delcypher, @MartinNowack)

benchmarks/graphviz/graphviz-2.34.0/lib/graph/graph.c
/* $Id$ $Revision$ */
/* vim:set shiftwidth=4 ts=8: */

/*************************************************************************
 * Copyright (c) 2011 AT&T Intellectual Property 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors: See CVS logs. Details at http://www.graphviz.org/
 *************************************************************************/

#include <limits.h>

#include "libgraph.h"

#ifdef DMALLOC
#include "dmalloc.h"
#endif

Dtdisc_t agNamedisc = {
    offsetof(Agnode_t, name),
    -1,
    -1,				/* link offset */
    NIL(Dtmake_f),
    NIL(Dtfree_f),
    NIL(Dtcompar_f),		/* use strcmp */
    NIL(Dthash_f),
    NIL(Dtmemory_f),
    NIL(Dtevent_f)
};

Dtdisc_t agNodedisc = {
    offsetof(Agnode_t, id),
    sizeof(int),
    -1,				/* link offset */
    NIL(Dtmake_f),
    NIL(Dtfree_f),
    (Dtcompar_f) agcmpid,
    NIL(Dthash_f),
    NIL(Dtmemory_f),
    NIL(Dtevent_f)
};

Dtdisc_t agIndisc = {
    0,				/* pass whole object as key */
    0,
    -1,				/* link offset */
    NIL(Dtmake_f),
    NIL(Dtfree_f),
    (Dtcompar_f) agcmpin,
    NIL(Dthash_f),
    NIL(Dtmemory_f),
    NIL(Dtevent_f)
};

Dtdisc_t agOutdisc = {
    0,				/* pass whole object as key */
    0,
    -1,				/* link offset */
    (Dtmake_f) 0,
    (Dtfree_f) 0,
    (Dtcompar_f) agcmpout,
    (Dthash_f) 0,
    (Dtmemory_f) 0,
    (Dtevent_f) 0
};

int agcmpid(Dt_t * dict, int *id0, int *id1, Dtdisc_t * disc)
{
    return (*id0) - *(id1);
}

#ifdef DEBUG
static int myinedgecmp(e0, e1)
Agedge_t *e0, *e1;
{
    int rv = myinedgecmp(e0, e1);
    printf("compare (%s,%s:%s),(%s,%s:%s) = %d\n",
	   e0->head ? e0->head->name : "nil",
	   e0->tail ? e0->tail->name : "nil",
	   e0->attr && e0->attr[KEYX]? e0->attr[KEYX] : "nil",
	   e1->head ? e1->head->name : "nil",
	   e1->tail ? e1->tail->name : "nil",
	   e1->attr && e1->attr[KEYX]? e1->attr[KEYX] : "nil",
	   rv);
    return rv;
}
#endif

static int keycmp(Agedge_t * e0, Agedge_t * e1)
{
    char *key0, *key1;
    key0 = e0->attr ? e0->attr[KEYX] : NULL;
    key1 = e1->attr ? e1->attr[KEYX] : NULL;
    if (key0 == NULL)
	return (key1 ? -1 : 0);
    if (key1 == NULL)
	return 1;
    return strcmp(key0, key1);
}

int agcmpin(Dict_t * d, Agedge_t * e0, Agedge_t * e1, Dtdisc_t * disc)
{
    int e0tailid, e0headid, e1tailid, e1headid;

    e0tailid = e0->tail ? e0->tail->id : -1;
    e0headid = e0->head ? e0->head->id : -1;
    e1tailid = e1->tail ? e1->tail->id : -1;
    e1headid = e1->head ? e1->head->id : -1;

    if (e0headid != e1headid)
	return e0headid - e1headid;
    if (e0tailid != e1tailid)
	return e0tailid - e1tailid;
    return keycmp(e0, e1);
}

int agcmpout(Dict_t * d, Agedge_t * e0, Agedge_t * e1, Dtdisc_t * disc)
{
    int e0tailid, e0headid, e1tailid, e1headid;

    e0tailid = e0->tail ? e0->tail->id : -1;
    e0headid = e0->head ? e0->head->id : -1;
    e1tailid = e1->tail ? e1->tail->id : -1;
    e1headid = e1->head ? e1->head->id : -1;

    if (e0tailid != e1tailid)
	return e0tailid - e1tailid;
    if (e0headid != e1headid)
	return e0headid - e1headid;
    return keycmp(e0, e1);
}

static Agdata_t *agnewdata(void)
{
    Agdata_t *rv;

    rv = NEW(Agdata_t);
    rv->node_dict = dtopen(&agNamedisc, Dttree);
    rv->globattr = agNEWdict("graph");
    rv->nodeattr = agNEWdict("node");
    rv->edgeattr = agNEWdict("edge");
    if (AG.proto_g) {
	agcopydict(rv->globattr, AG.proto_g->univ->globattr);
	agcopydict(rv->nodeattr, AG.proto_g->univ->nodeattr);
	agcopydict(rv->edgeattr, AG.proto_g->univ->edgeattr);
    }
    return rv;
}

static void agfreedata(Agraph_t * g)
{
    agFREEdict(g, g->univ->globattr);
    agFREEdict(g, g->univ->nodeattr);
    agFREEdict(g, g->univ->edgeattr);
    dtclose(g->univ->node_dict);
    free(g->univ);
}

static void dup_proto(Agraph_t * g, Agproto_t * proto)
{
    Agnode_t *n = NULL;
    Agedge_t *e = NULL;
    Agproto_t *s = NEW(Agproto_t);

    s->prev = g->proto;
    if (proto) {
	n = proto->n;
	e = proto->e;
    }
    s->n = agNEWnode(g, "\001proto", n);
    s->e = agNEWedge(g, s->n, s->n, e);
    g->proto = s;
}

void agpushproto(Agraph_t * g)
{
    dup_proto(g, g->proto);
}

void agpopproto(Agraph_t * g)
{
    Agproto_t *s = g->proto;
    if (s != NULL) {
	g->proto = s->prev;
	s->e->tail = s->e->head = s->n;
	agFREEedge(s->e);
	agFREEnode(s->n);
	free(s);
    }
}

static Agraph_t *agNEWgraph(char *name, Agraph_t * parent, int kind)
{
    int i, nobj;
    Agraph_t *g;

    if (AG.init_called == FALSE) {
	agerr(AGERR, "libag error -- aginit() was not called\n");
	return 0;
    }
    g = (Agraph_t *) calloc(1, AG.graph_nbytes);
    g->tag = TAG_GRAPH;
    g->kind = kind;
    g->nodes = dtopen(&agNodedisc, Dttree);
    g->inedges = dtopen(&agIndisc, Dttree);
    g->outedges = dtopen(&agOutdisc, Dttree);

    if (parent == NULL) {
	g->univ = agnewdata();
	g->root = g;
	nobj = dtsize(g->univ->globattr->dict);
	if (nobj) {
	    g->attr = N_NEW(nobj, char *);
		g->didset = N_NEW((nobj + CHAR_BIT - 1) / CHAR_BIT, char);
	}
	else {
	    g->attr = NULL;
		g->didset = NULL;
	}
	for (i = 0; i < nobj; i++)
	    g->attr[i] = agstrdup(AG.proto_g->attr[i]);
	} else {
	g->univ = parent->univ;
	g->root = parent->root;
	nobj = dtsize(parent->univ->globattr->dict);
	if (nobj) {
	    g->attr = N_NEW(nobj, char *);
		g->didset = N_NEW((nobj + CHAR_BIT - 1) / CHAR_BIT, char);
	}
	else {
	    g->attr = NULL;
		g->didset = NULL;
	}
	for (i = 0; i < nobj; i++)
	    g->attr[i] = agstrdup(parent->attr[i]);

    }

    g->meta_node = NULL;
    g->name = agstrdup(name);
    g->proto = NULL;

    if (parent)
	dup_proto(g, parent->proto);
    else
	agpushproto(g);
    return g;
}

static int reach0(Dict_t * m, Agnode_t * from, Agnode_t * to)
{
    Agedge_t *e;

    if (from == to)
	return TRUE;
    if (agfindedge(from->graph->root, from, to))
	return TRUE;
    dtinsert(m, from);
    for (e = agfstout(from->graph, from); e; e = agnxtout(from->graph, e))
	if ((dtsearch(m, e->head) == NULL) && reach0(m, e->head, to))
	    return TRUE;
    return FALSE;
}

static int reach(Agnode_t * from, Agnode_t * to)
{
    Dict_t *m;
    int rv;

    m = dtopen(&agNodedisc, Dttree);
    rv = reach0(m, from, to);
    dtclose(m);
    return rv;
}

Agraph_t *agusergraph(Agnode_t * n)
{
    return (n->graph->meta_node ? NULL : (Agraph_t *) (n->attr[0]));
}




Agraph_t *agopen(char *name, int kind)
{
    Agraph_t *g, *meta;

    g = agNEWgraph(name, NULL, kind);
    meta = agNEWgraph(name, NULL, AGMETAGRAPH);
    if (!g || !meta)
	return 0;
    agnodeattr(meta, "agusergraph", NULL);
    g->meta_node = agnode(meta, name);
    g->meta_node->attr[0] = (char *) g;
    return g;
}

Agraph_t *agsubg(Agraph_t * g, char *name)
{
    Agraph_t *subg, *meta;
    Agnode_t *n;

    meta = g->meta_node->graph;
    n = agfindnode(meta, name);
    if (n)
	subg = agusergraph(n);
    else {
	subg = agNEWgraph(name, g, g->kind);
	if (!subg)
	    return 0;
	n = agnode(meta, name);
	subg->meta_node = n;
	n->attr[0] = (char *) subg;
    }
    agINSgraph(g, subg);
    return subg;
}

Agraph_t *agfindsubg(Agraph_t * g, char *name)
{
    Agnode_t *n;

    if (g->meta_node) {
	n = agfindnode(g->meta_node->graph, name);
	if (n)
	    return agusergraph(n);
    }
    return NULL;
}

void agINSgraph(Agraph_t * g, Agraph_t * subg)
{
    Agnode_t *h, *t;
    t = g->meta_node;
    h = subg->meta_node;
    if (t && h && (reach(h, t) == FALSE))
	agedge(t->graph, t, h);
}

void agclose(Agraph_t * g)
{
    Agedge_t *e, *f;
    Agnode_t *n, *nn;
    Agraph_t *meta = NULL;
    int i, nobj, flag, is_meta;

    if ((g == NULL) || (TAG_OF(g) != TAG_GRAPH))
	return;
    is_meta = AG_IS_METAGRAPH(g);
    if (is_meta == FALSE) {
	meta = g->meta_node->graph;
	/* recursively remove its subgraphs */
	do {			/* better semantics would be to find strong component */
	    flag = FALSE;
	    for (e = agfstout(meta, g->meta_node); e; e = f) {
		f = agnxtout(meta, e);
		if (agnxtin(meta, agfstin(meta, e->head)) == NULL) {
		    agclose(agusergraph(e->head));
		    flag = TRUE;
		}
	    }
	} while (flag);
    }
    while (g->proto)
	agpopproto(g);
    if (is_meta == FALSE) {
	nobj = dtsize(g->univ->globattr->dict);
	for (i = 0; i < nobj; i++)
	    agstrfree(g->attr[i]);
    }
    if (g->attr)
	free(g->attr);
	if (g->didset)
		free(g->didset);
    if (g == g->root) {
	for (n = agfstnode(g); n; n = nn) {
	    nn = agnxtnode(g, n);
	    agDELnode(g, n);
	}
	if (is_meta == FALSE)
	    agclose(g->meta_node->graph);
	agfreedata(g);
    } else {
	if (is_meta == FALSE)
	    agdelete(meta, g->meta_node);
    }
    dtclose(g->nodes);
    dtclose(g->inedges);
    dtclose(g->outedges);
    agstrfree(g->name);
    TAG_OF(g) = -1;
    free(g);
}

int agcontains(Agraph_t * g, void *obj)
{
    switch (TAG_OF(obj)) {
    case TAG_NODE:
	return (agidnode(g, ((Agnode_t *) obj)->id) != NULL);
    case TAG_EDGE:
	return (dtsearch(g->inedges, (Agedge_t *) obj) != NULL);
    case TAG_GRAPH:
	return (reach(g->meta_node, ((Agraph_t *) obj)->meta_node));
    }
    return FALSE;
}

void aginsert(Agraph_t * g, void *obj)
{
    switch (TAG_OF(obj)) {
    case TAG_NODE:
	agINSnode(g, (Agnode_t*)obj);
	break;
    case TAG_EDGE:
	agINSedge(g, (Agedge_t*)obj);
	break;
    case TAG_GRAPH:
	agINSgraph(g, (Agraph_t*)obj);
	break;
    }
}

void agdelete(Agraph_t * g, void *obj)
{
    switch (TAG_OF(obj)) {
    case TAG_NODE:
	agDELnode(g, (Agnode_t*)obj);
	break;
    case TAG_EDGE:
	agDELedge(g, (Agedge_t*)obj);
	break;
    case TAG_GRAPH:
	agclose((Agraph_t*)obj);
	break;
    }
}

int agnnodes(Agraph_t * g)
{
    return dtsize(g->nodes);
}

int agnedges(Agraph_t * g)
{
    return dtsize(g->outedges);
}
